### PR TITLE
Fix AddGuildMemberAsync

### PR DESCRIPTION
**Pull request by Plerx2493**
# Summary
- NRE: Discord returns 204 No content without a body if the user is already in the guild
- KeyNotFoundException: DiscordUser wasnt added to the membercache

### CODE REVIEW

File: DSharpPlus.Rest/DiscordRestClient.cs [hunk 1]
```diff
@@ -376,8 +376,8 @@ public async Task LeaveGuildAsync(ulong guild_id)
     /// <param name="roles">User roles</param>
     /// <param name="muted">Whether this user should be muted on join</param>
     /// <param name="deafened">Whether this user should be deafened on join</param>
-    /// <returns></returns>
-    public async Task<DiscordMember> AddGuildMemberAsync(ulong guild_id, ulong user_id, string access_token, string nick, IEnumerable<DiscordRole> roles, bool muted, bool deafened)
+    /// <returns>Only returns the member if the was not already in the guild</returns>
+    public async Task<DiscordMember?> AddGuildMemberAsync(ulong guild_id, ulong user_id, string access_token, string nick, IEnumerable<DiscordRole> roles, bool muted, bool deafened)
```
Comment (akiraveliara) [inline DSharpPlus.Rest/DiscordRestClient.cs:380]: "if they were not"/"if the member was not"
also, this sounds like a good opportunity to fix the parameter naming

File: DSharpPlus/Entities/Guild/DiscordGuild.cs [hunk 2]
```diff
@@ -813,12 +813,12 @@ public async Task<IReadOnlyList<DiscordMember>> SearchMembersAsync(string name,
...
     /// <param name="muted">whether this user has to be muted</param>
     /// <param name="deaf">whether this user has to be deafened</param>
-    /// <returns></returns>
+    /// <returns>Only returns the member if it was not already in the guild</returns>
     /// <exception cref="UnauthorizedException">Thrown when the client does not have the <see cref="Permissions.CreateInstantInvite" /> permission.</exception>
     /// <exception cref="NotFoundException">Thrown when the <paramref name="user"/> or <paramref name="access_token"/> is not found.</exception>
     /// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
     /// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
-    public async Task AddMemberAsync(DiscordUser user, string access_token, string nickname = null, IEnumerable<DiscordRole> roles = null,
+    public async Task<DiscordMember?> AddMemberAsync(DiscordUser user, string access_token, string nickname = null, IEnumerable<DiscordRole> roles = null,
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Guild/DiscordGuild.cs:821]: same point applies here

File: DSharpPlus/Net/Rest/DiscordApiClient.cs [hunk 3]
```diff
@@ -613,8 +613,18 @@ ulong guildId
...
+        {
+            // User was already in the guild, Discord doesn't return the member object in this case
+            return null;
+        }
 
         TransportMember transport = JsonConvert.DeserializeObject<TransportMember>(res.Response!)!;
+        
+        DiscordUser usr = new(transport.User) { Discord = this._discord! };
+
+        this._discord!.UpdateUserCache(usr);
```
Comment (akiraveliara) [inline DSharpPlus/Net/Rest/DiscordApiClient.cs:627]: are we sure we always have a discord client through oauth? worth testing

**Review by akiraveliara — CHANGES_REQUESTED** [review]

File: DSharpPlus/Net/Rest/DiscordApiClient.cs (hunk 3 above)
Comment (Plerx2493) [reply DSharpPlus/Net/Rest/DiscordApiClient.cs:627]: UserCache is on `BaseDiscordClient`

**Review by akiraveliara — APPROVED** [review]

**Pull request merged**